### ISSUE TITLE
perldebguts documentation should be fixed

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16055]
Migrated from [rt.perl.org#131672](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=131672) (status was 'resolved')

Searchable as RT131672$

**Comment by p5pRT** [comment]
### From @KES777

#### Created by @KES777



\[The documentation\]\(https://perldoc.perl.org/perldebguts.html)
should be fixed&#8203;:

\-\-\- \(This doesn't happen if the subroutine was compiled in the DB package\.\)
\+\+\+ \(This doesn't happen if the subroutine was called from DB package\.\)


In short&#8203;:
Despite on 'from\_db' sub is compiled at main&#8203;:: package we do NOT see calls to 'DB&#8203;::sub' for it
Becase CALL to this sub is made from DB&#8203;:: package

And vice versa&#8203;:
Despite on 'DB&#8203;::test' is compiled at DB&#8203;:: package we see calls to 'DB&#8203;::sub' for it
Becase CALL to this sub is made not from DB&#8203;:: package

The output&#8203;:
$ perl \-Ilib \-d&#8203;:DB test\.pl
Called&#8203;: main&#8203;::BEGIN
Called&#8203;: DB&#8203;::test


$cat test\.pl
sub from\_db \{\};
use Test;
DB&#8203;::test\(\);


$cat lib/Test\.pm
package Test;
1;

$cat lib/Devel/DB\.pm
package DB;

use B qw/ svref\_2object /;
sub sub\_name \{
&nbsp;    my $r = shift;
&nbsp;    return $r unless ref $r;
&nbsp;    return $r unless my $cv =  svref\_2object\( $r \);
&nbsp;    return $r unless $cv\->isa\( 'B&#8203;::CV' \)
&nbsp;              and my $gv = $cv\->GV
&nbsp;              ;
&nbsp;    my $name = '';
&nbsp;    if \( my $st = $gv\->STASH \) \{
&nbsp;        $name = $st\->NAME \. '&#8203;::';
&nbsp;    \}
&nbsp;    my $n = $gv\->NAME;
&nbsp;    if \( $n \) \{
&nbsp;        $name \.= $n;
&nbsp;        if \( $n eq '\_\_ANON\_\_' \) \{
&nbsp;            $name \.= ' defined at ' \. $gv\->FILE \. '&#8203;:' \. $gv\->LINE;
&nbsp;        \}
&nbsp;    \}
&nbsp;    return $name;
\}


sub test\{\};

sub postponed \{
&nbsp;        \# Call only after DB&#8203;:: is loaded
&nbsp;        main&#8203;::from\_db\(\)   if shift ne '\*main&#8203;::\_\<lib/Devel/DB\.pm';
\}

sub DB\{
&nbsp;        main&#8203;::from\_db\(\);
\}

sub sub \{
&nbsp;        print "Called&#8203;: "\, sub\_name\( $DB&#8203;::sub \)\, "\\n";
&nbsp;        main&#8203;::from\_db\(\);

&nbsp;        &$DB&#8203;::sub;
\}

1;


See attached archive with source codes\.

<details><summary>Perl Info</summary>

```

Flags:
    category=docs
    severity=low

Site configuration information for perl 5.24.1:

Configured by kes at Sun Apr 30 22:40:30 EEST 2017.

Summary of my perl5 (revision 5 version 24 subversion 1) configuration:
   
  Platform:
    osname=linux, osvers=4.4.0-53-generic, archname=x86_64-linux
    uname='linux kes-x751sa 4.4.0-53-generic #74-ubuntu smp fri dec 2 15:59:10 utc 2016 x86_64 x86_64 x86_64 gnulinux '
    config_args='-de -Dprefix=/home/kes/perl5/perlbrew/perls/perl-5.24.1 -Aeval:scriptdir=/home/kes/perl5/perlbrew/perls/perl-5.24.1/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion='', gccversion='5.4.0 20160609', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.23.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.23'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'

Locally applied patches:
    Devel::PatchPerl 1.38


@INC for perl 5.24.1:
    /home/kes/perl5/perlbrew/perls/perl-5.24.1/lib/site_perl/5.24.1/x86_64-linux
    /home/kes/perl5/perlbrew/perls/perl-5.24.1/lib/site_perl/5.24.1
    /home/kes/perl5/perlbrew/perls/perl-5.24.1/lib/5.24.1/x86_64-linux
    /home/kes/perl5/perlbrew/perls/perl-5.24.1/lib/5.24.1


Environment for perl 5.24.1:
    HOME=/home/kes
    LANG=en_US.UTF-8
    LANGUAGE=en
    LC_ADDRESS=uk_UA.UTF-8
    LC_IDENTIFICATION=uk_UA.UTF-8
    LC_MEASUREMENT=uk_UA.UTF-8
    LC_MESSAGES=en_US.UTF-8
    LC_MONETARY=uk_UA.UTF-8
    LC_NAME=uk_UA.UTF-8
    LC_NUMERIC=uk_UA.UTF-8
    LC_PAPER=uk_UA.UTF-8
    LC_TELEPHONE=uk_UA.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/kes/perl5/perlbrew/bin:/home/kes/perl5/perlbrew/perls/perl-5.24.1/bin:/home/kes/bin:/home/kes/bin:/home/kes/bin:/home/kes/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
    PERLBREW=command perlbrew
    PERLBREW_BASHRC_VERSION=0.78
    PERLBREW_HOME=/home/kes/.perlbrew
    PERLBREW_MANPATH=/home/kes/perl5/perlbrew/perls/perl-5.24.1/man
    PERLBREW_PATH=/home/kes/perl5/perlbrew/bin:/home/kes/perl5/perlbrew/perls/perl-5.24.1/bin
    PERLBREW_PERL=perl-5.24.1
    PERLBREW_ROOT=/home/kes/perl5/perlbrew
    PERLBREW_VERSION=0.78
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @KES777

[bugreport.tar](https://rt-archive.perl.org/perl5/Ticket/Attachment/1470261/797040/bugreport.tar)

**Comment by p5pRT** [comment]
### From zefram@fysh.org

KES wrote&#8203;:
>\-\-\- \(This doesn't happen if the subroutine was compiled in the DB package\.\)
>\+\+\+ \(This doesn't happen if the subroutine was called from DB package\.\)

The phrase "the subroutine" is ambiguous\, because in that paragraph
there are two subroutines in play\.  The original could be clarified
by expanding to "the calling subroutine"\.  Your proposed change would
retain the ambiguity\.

\-zefram

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @KES777


> KES wrote&#8203;:
>> \-\-\- \(This doesn't happen if the subroutine was compiled in the DB package\.\)
>> \+\+\+ \(This doesn't happen if the subroutine was called from DB package\.\)

Actually first is right\. Because with current behavior&#8203;:
&nbsp;  when a call to &DB&#8203;::sub is not happen if the subroutine was called from DB package

I can not setup breakpoint by subroutine name\.

I  check $DB&#8203;::sub and setup breakpoint at 'DB&#8203;::sub'\. Code looks like next&#8203;:

&nbsp;    if\( $breakpoints\{ $DB&#8203;::sub \} \) \{ $DB&#8203;::single = 1 \};

This works well in most cases except for next one\.

I implement debugger command in not DB package\. 
I do this because want to be able to debug debugger command\.

So when that subroutine \(which implements debugger command\) is called from DB package\, it can not be tracked\.
&nbsp;   A call to &DB&#8203;::sub is not happen

As work around I put next code into that sub&#8203;:

&nbsp;    $^D |= \(1\<\<30\); $DB&#8203;::single = 1; 

But this will not be required if perl works as documented&#8203;:
&nbsp;     A call to &DB&#8203;::sub is happened for all subroutines which was not compiled in the DB package\.
&nbsp;     despite on place of call\.


I check this behavior on the oldest version I have&#8203;: perl\-5\.6\.2
it works contrary to the documentation too =\(

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Documentation fixed in commit 7270698f29ae2f22f2cdfec7b156a75602cddf68\.

\-zefram

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @KES777

On Tue\, 05 Dec 2017 10&#8203;:11&#8203;:53 \-0800\, zefram@&#8203;fysh\.org wrote&#8203;:
> Documentation fixed in commit 7270698f29ae2f22f2cdfec7b156a75602cddf68\.
> 
> \-zefram

Thank you for fix\.

Zefram\, will next sound more clear? Please review\.
When execution of the program reaches a subroutine call\, a call to &DB&#8203;::sub \(args\) is made instead\, with $DB&#8203;::sub holding the name of the called subroutine\. \(This happen for all debuggable subroutines\. Subroutines compiled in DB&#8203;:: package all are not debuggable\.\)

**Comment by p5pRT** [comment]
### From zefram@fysh.org

KES via RT wrote&#8203;:
>Zefram\, will next sound more clear?

No\, that's less clear\.

If you want to raise another issue with the documentation\, please do so
in a separate ticket\.

\-zefram

**Comment by p5pRT** [comment]
### From @khwilliamson

Thank you for filing this report\.  You have helped make Perl better\.

With the release yesterday of Perl 5\.28\.0\, this and 185 other issues have been
resolved\.

Perl 5\.28\.0 may be downloaded via&#8203;:
https://metacpan.org/release/XSAWYERX/perl-5.28.0

If you find that the problem persists\, feel free to reopen this ticket\.

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
